Bayesian Binary Sensor

jlmcgehee21 opened the pull request:
## Description:
Why:

* It would be beneficial to leverage various sensor outputs in a
Bayesian manner in order to sense more complex events.

This change addresses the need by:

* `BayesianBinarySensor` class in
`./homeassistant/components/binary_sensor/bayesian.py`
* Tests in `./tests/components/binary_sensor/test_bayesian.py`

Caveats:
This is my first time in this code-base. I did try to follow conventions
that I was able to find, but I'm sure there will be some issues to
straighten out.

## Example entry for `configuration.yaml` (if applicable):
```
binary_sensor:
  name: 'currently_cooking'
  platform: 'bayesian'
  prior: 0.1
  probability_threshold: 0.7
  observations:
    - entity_id: 'switch.kitchen_lights'
      prob_given_true: 0.6
      prob_given_false: 0.2
      platform: 'state'
      to_state: 'on'
    - entity_id: 'sensor.stove_temperature'
      prob_given_true: 0.9
      platform: 'numeric_state'
      above: 100
    - entity_id: 'sensor.kitchen_motion'
      prob_given_true: 0.5
      prob_given_false: 0.2
      platform: 'state'
      to_state: 'on'

binary_sensor:
  name: 'in_bed'
  platform: 'bayesian'
  prior: 0.5
  probability_threshold: 0.95
  observations:
    - entity_id: 'sensor.living_room_motion'
      prob_given_true: 0.4
      prob_given_false: 0.2
      platform: 'state'
      to_state: 'off'
    - entity_id: 'sensor.basement_motion'
      prob_given_true: 0.5
      prob_given_false: 0.4
      platform: 'state'
      to_state: 'off'
    - entity_id: 'sensor.bedroom_motion'
      prob_given_true: 0.5
      platform: 'state'
      to_state: 'on'
    - entity_id: 'sensor.sun'
      prob_given_true: 0.7
      platform: 'state'
      to_state: 'below_horizon'
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)  https://github.com/home-assistant/home-assistant.github.io/pull/3268


If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jlmcgehee21 committed:
Bayesian Binary Sensor

Why:

* It would be beneficial to leverage various sensor outputs in a
Bayesian manner in order to sense more complex events.

This change addresses the need by:

* `BayesianBinarySensor` class in
`./homeassistant/components/binary_sensor/bayesian.py`
* Tests in `./tests/components/binary_sensor/test_bayesian.py`

Caveats:
This is my first time in this code-base. I did try to follow conventions
that I was able to find, but I'm sure there will be some issues to
straighten out.

homeassistant commented:
Hi @jlmcgehee21, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8810).

Once you do that we will be able to review and accept this pull request.

Thanks!

jlmcgehee21 committed:
minor cleanup

jlmcgehee21 commented:
Was running tests locally with `ptw tests/components/binary_sensor homeassistant/components/binary_sensor/ -- -x tests/components/binary_sensor/test_bayesian.py` and had no trouble getting them to pass in isolation.

I probably will need some pointers regarding whatever side effects I may be causing when running the full test suite.

arsaboo commented:
Love it....will be great for presence detection as well.

cmsimike commented:
This will be fantastic to try to infer when my apartment should go into night/sleep mode!

Danielhiversen commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,174 @@
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.binary_sensor import (BinarySensorDevice,
+                                                    PLATFORM_SCHEMA)
+from homeassistant.const import (CONF_NAME, STATE_UNKNOWN, CONF_SENSOR_CLASS,
```
CONF_SENSOR_CLASS does not exists in `homeassistant.const`

Danielhiversen commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,174 @@
+        else:
+            self.current_obs.pop(entity, None)
+
+    def __update_probability(self, prior, observation):
```
You are not using `self` in the calculations, so the function could be extracted from the class.
We normally use only one `_` in function name,

jlmcgehee21 replied:
Oops, was updating `self.probability` before a refactor.  Changed to a `@staticmethod` to keep the namespace of the class (my usual approach), if this is not preferred, I can move outside.

jlmcgehee21 committed:
Address reviewer's comments

This change addresses the need by:

* Removing `CONF_SENSOR_CLASS` and its usage in `get_deprecated`.
* Make probability update function a static method, and use single `_`
to match project conventions.

jlmcgehee21 committed:
Address linter failures

Danielhiversen commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+    observations = config.get(CONF_OBSERVATIONS)
+    prior = config.get(CONF_PRIOR)
+    probability_threshold = config.get(CONF_PROBABILITY_THRESHOLD)
+    device_class = CONF_DEVICE_CLASS
```
Should this be `config.get(CONF_DEVICE_CLASS)` ?

jlmcgehee21 committed:
fix `device_class` declaration

jlmcgehee21 commented:
@Danielhiversen, once I add docs in home-assistant.github.io, will this make the next release?

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+
+        for obs in self._observations:
+            entity_id = obs['entity_id']
+            async_track_state_change(hass, entity_id,
```
Do this in the method `async_added_to_hass` which will be called when the entity is added to home assistant.

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+class BayesianBinarySensor(BinarySensorDevice):
+    """Representation of a Bayesian sensor."""
+
+    def __init__(self, hass, name, prior, observations, probability_threshold,
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+                                         entity_observation.get('below'),
+                                         entity_observation.get('above'), None,
+                                         entity_observation):
+
```
Remove blank line.

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Get the latest data and updates the states."""
```
Write all the verbs in imperative mood: "Get... update..."

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+Use Bayesian Inference to trigger a binary sensor.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.bayesian_binary/
```
`https://home-assistant.io/components/binary_sensor.bayesian/`

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+
+            prior = self.prior
+            for obs in self.current_obs.values():
+                prior = self._update_probability(obs, prior)
```
The signature of the `_update_probability` method has the parameters in the reverse order of this call. Is this deliberate? I'm not very familiar with the Bayesian theory in application.

jlmcgehee21 replied:
Good catch! It actually didn't matter (and hence didn't fail my tests) because I was treating the scenario a bit naively (`P(Event|False) = 1 - P(Event|True)`).  I thought this would be a decent first shot, but after some reflection, I preferred to allow the user to specify `P(Event|False)` and default to the above case if they do not.

MartinHjelmare reviewed (changes requested):
Cool! Some comments below.

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,173 @@
+    observations = config.get(CONF_OBSERVATIONS)
+    prior = config.get(CONF_PRIOR)
+    probability_threshold = config.get(CONF_PROBABILITY_THRESHOLD)
+    device_class = config.get(CONF_DEVICE_CLASS)
```
`CONF_DEVICE_CLASS` isn't part of the config schema.

jlmcgehee21 committed:
Address Comments

Why:
* Not validating config schema enough.
* Not following common practices for async initialization.
* Naive implementation of Bayes' rule.

This change addresses the need by:
* Improving config validation for observations.
* Moving initialization logic into `async_added_to_hass`.
* Re-configuring Bayesian updates to allow true P|Q usage.

jlmcgehee21 committed:
address linting issues

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,225 @@
+    vol.Coerce(float),
+    vol.Optional(CONF_P_GIVEN_F):
+    vol.Coerce(float)
+})
```
You can set all the nodes to be required in the schema except the ones marked as optional with the `required` keyword argument.
```py
vol.Schema({...}, required=True)
```

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -92,51 +133,59 @@ def async_threshold_sensor_state_listener(entity, old_state,
             self.watchers[platform](entity_obs)
 
             prior = self.prior
+            print(self.current_obs.values())
```
Remove debug print.

MartinHjelmare commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -92,51 +133,59 @@ def async_threshold_sensor_state_listener(entity, old_state,
+            entity_observation.get('below'),
+            entity_observation.get('above'), None, entity_observation)
+
+        if should_trigger:
```
The code between the `if` and `else` here makes a pattern that is repeated for both observation platforms. Maybe break it out into a helper method?

jlmcgehee21 committed:
Improve DRYness by adding `_update_current_obs` method

MartinHjelmare commented:
Fix the two lint issues and this is good to go.

pvizeli commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,217 @@
+            async_track_state_change(self.hass, entity_id,
+                                     async_threshold_sensor_state_listener)
+
+    def _update_current_obs(self, entity_observation, should_trigger):
```
add a function documentation. Add decorator "@callback" if they need run inside async loop or add "Async friendly" to function documentation if that dosn't matters.

jlmcgehee21 replied:
I basically adapted the `binary.threshold` sensor and its test for this implementation, and it seems the test is made to run in what looks like a full hass environment.  With this in mind, I would guess that the `@callback` decorator is not needed because my tests are passing? If this is true, then appropriate solution would be to add "Async friendly" to the docstring, correct?

As a side note, the `pylintrc` and the travis tests both designate docstrings are only needed for public methods. For ease of contribution, would it be appropriate to include modification of these files as part of my PR in order for linting to fail if these methods have no documentation? 

I appreciate the feedback, and hope to address it the best way possible.

pvizeli replied:
Good point. I make a PR to remove that from `pylintrc`. This small info about that function is in every case helpfull.

Yeah, the function can run async/sync. So but "Async friendly" to this function like https://github.com/home-assistant/home-assistant/blob/c56f99baafac33483dd13699993d7da4ee5d7efd/homeassistant/helpers/location.py#L11-L14.

That help to prevent on future change on that platform.

jlmcgehee21 replied:
:+1: thanks for the clarification.

pvizeli commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,217 @@
+        else:
+            self.current_obs.pop(entity, None)
+
+    def _process_numeric_state(self, entity_observation):
```
add a function documentation. Add decorator "@callback" if they need run inside async loop or add "Async friendly" to function documentation if that dosn't matters.

pvizeli commented on homeassistant/components/binary_sensor/bayesian.py:
```diff
@@ -0,0 +1,217 @@
+
+        self._update_current_obs(entity_observation, should_trigger)
+
+    def _process_state(self, entity_observation):
```
add a function documentation. Add decorator "@callback" if they need run inside async loop or add "Async friendly" to function documentation if that dosn't matters.

pvizeli reviewed (changes requested):
Looks good

jlmcgehee21 committed:
update doc strings and ensure functions are set up properly for async

jlmcgehee21 commented:
@Danielhiversen @MartinHjelmare @pvizeli hopefully I have addressed all comments/concerns.  I have also added documentation, and linked to the PR in the description for this PR.  Unfortunately, it seems I missed the `0.52` release, but hopefully we can make the next one!

MartinHjelmare reviewed (approved):
~~I'm not sure about the callback annotations, but the rest looks good.~~
Everything is good.

pvizeli committed:
Make only 1 state change handle

pvizeli committed:
fix style

pvizeli committed:
fix style part 2

pvizeli committed:
fix lint

pvizeli commented:
Retrigger travis API

pvizeli commented:
Travis pass: https://travis-ci.org/home-assistant/home-assistant/builds/269760101

moskovskiy82 commented:
Will this make into 0.53?

MartinHjelmare commented:
Yes.
